src/freenet/node/DarknetPeerNode.java

```diff
@@ -1466,19 +1477,97 @@ public int sendFileOffer(HTTPUploadedFile file, String message) throws IOExcepti
 		return sendFileOffer(fnam, mime, message, data);
 	}
 
+	// handler for sendTextFeed
 	public void handleFproxyN2NTM(SimpleFieldSet fs, int fileNumber) {
 		String text = null;
 		long composedTime = fs.getLong("composedTime", -1);
 		long sentTime = fs.getLong("sentTime", -1);
 		long receivedTime = fs.getLong("receivedTime", -1);
-	  	try {
+		try {
 			text = Base64.decodeUTF8(fs.get("text"));
 		} catch (IllegalBase64Exception e) {
 			Logger.error(this, "Bad Base64 encoding when decoding a N2NTM SimpleFieldSet", e);
 			return;
 		}
-		N2NTMUserAlert userAlert = new N2NTMUserAlert(this, text, fileNumber, composedTime, sentTime, receivedTime);
+		long msgid = fs.getLong("msgid", -1);
+		String newText = text;
+		List<UserAlert> merged = new ArrayList<>();
+		// TODO: extract merging logic into private method
+		if (msgid != -1) {
+			// merge existing user alerts
+
+			// TODO: This costs time linear to the number
+			// of known alerts, so receiving a large text
+			// message is quadratic in its size! Worst
+			// case cost for transfer of a single huge
+			// n2ntm (when receiving every second message)
+			// is about 0.5 N*N, with N at most
+			// 128. However even sending a 64kiB file
+			// (DarknetPeerNode.java) does not cause
+			// noticeable load on the receiving system.
+			synchronized (node.clientCore.alerts) {
```

Comment:
Now that you have partIndex numbers this should use them for assembling the message by putting the parts into a String[] with partIndex being the offset, and then squashing the array into a single String.